mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -2096,12 +2118,50 @@ populateHostHarnessLogic(ModuleOp &module,
       printp = printResults.getValue();
     if (printp)
       emitPrintTensor(b, lvar);
-    // dealloc lvar
+  }
+
+  for (auto &vvar : valVars) {
+    b.create<memref::DeallocOp>(loc, vvar);
+  }
+  for (auto &lvar : localVars) {
     b.create<memref::DeallocOp>(loc, lvar);
   }
 
   b.create<func::ReturnOp>(loc, ValueRange{});
 
+  // Wrap the kernels and gather them to substitute in calls.
+  llvm::SmallDenseMap<func::FuncOp, func::FuncOp> wrappedFuncs;
+  for (auto &kernel : kernels) {
+    if (kernel.func->hasAttr("kernel")) {
+      wrappedFuncs[kernel.func] = createGPUWrapper(module, kernel);
+    } else {
+      wrappedFuncs[kernel.func] = kernel.func;
+    }
+  }
+
+  // Redirect calls to kernel functions to point at wrapped functions.
+  module.walk([&](Operation *op) -> WalkResult {
```

Comment:
Can be run on a CallOpInterface for both CallOp and async::LaunchOp

Reply:
Also simplified things.